### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by raszi**
Sorry for the huge PR! Design decisions are below.

## Faraday

I needed to implement the import lead feature, which needs a multpart POST request to be made to the `/bulk/v1/leads.json` endpoint. The standard [`Net::HTTP`](http://ruby-doc.org/stdlib-2.2.2/libdoc/net/http/rdoc/Net/HTTP.html) does not support multipart post so I needed to include the [`multipart-post`](https://rubygems.org/gems/multipart-post) but instead of including only that and stuck with the old `MktoRest::HttpUtils` I decided to include [`faraday`](https://rubygems.org/gems/faraday) because that provides a common interface for multiple adapters.

Authors of faraday are re-implementing everything in [`hurley`](https://rubygems.org/gems/hurley) see lostisland/faraday#454. But unfortunately it is not mature enough (v0.1) so I decided to depend on this widely used gem instead.

## Concerns

I extracted the concerns into separate files, so from now on we have:

 - connection building;
 - authentication;
 - CRUD lead;
 - import lead related concerns.

Any new functionality could be implemented in these and it will be easier to understand and would not bloat the client itself.

## Removing `Mkto::Lead` domain object

It did not give wide enough functionality and broke the common interface for other methods what this library provides. Some could implement a full-featured domain objects and create a common interface for that but sometimes returning a response sometimes an array of domain object makes no sense.

## Rewriting the tests

Having a common place for all the stubs may sound a good idea but reusing those stubs did not really help and just made the development harder. Maybe it would be a great idea to start using VCR to have real-world tests but for the sake of this PR I kept using [`webmock`](https://rubygems.org/gems/webmock).

The code-coverage for this PR is 100%, but I am not saying that it covers all the use cases and errors.

## Removing the examples directory

The specs should and do provide good enough examples. Maintaining two sets of samples makes no sense and hard to maintain.

## Includes these previous pull-requests

Resolves #7
Resolves #8
Resolves #11

### CODE REVIEW

**Comment by thenonameguy** [comment]
:+1:

File: .gitignore [hunk 1]
```diff
@@ -1,2 +1,9 @@
-.mktorest
-Gemfile.lock
```
Comment (raszi) [inline .gitignore]: It is a [good pattern](http://stackoverflow.com/questions/7517524/understanding-the-gemfile-lock-file) to commit in the `Gemfile.lock.

**Comment by jalemieux** [comment]
@raszi this is a lot better than the current implementation! It will also probably serve users out there better. 
What I suggest we do is (1) get your code in its own repo, and (2) update marketo doc to link to it for a prime example of using APIs with ruby. If you can take care of (1), i'll take care of (2).

**Comment by raszi** [comment]
[Forked](/raszi/mrkt) and [released](https://rubygems.org/gems/mrkt).

**Pull request closed without merging**